// @storybook/html/dist/entry-preview-docs.js
import "./_browser-chunks/chunk-4BE7D4DS.js";

// src/entry-preview-docs.ts
import { SourceType as SourceType2 } from "storybook/internal/docs-tools";

// src/docs/sourceDecorator.ts
import { SourceType } from "storybook/internal/docs-tools";
import { emitTransformCode, useEffect, useRef } from "storybook/preview-api";
function skipSourceRender(context) {
  let sourceParams = context?.parameters.docs?.source, isArgsStory = context?.parameters.__isArgsStory;
  return sourceParams?.type === SourceType.DYNAMIC ? !1 : !isArgsStory || sourceParams?.code || sourceParams?.type === SourceType.CODE;
}
var sourceDecorator = (storyFn, context) => {
  let source = useRef(void 0), story = storyFn();
  return useEffect(() => {
    let renderedForSource = context?.parameters.docs?.source?.excludeDecorators ? context.originalStoryFn(context.args, context) : story;
    skipSourceRender(context) || (typeof renderedForSource == "string" && source.current !== renderedForSource ? (emitTransformCode(renderedForSource, context), source.current = renderedForSource) : renderedForSource instanceof Element && source.current !== renderedForSource.outerHTML && (emitTransformCode(renderedForSource.outerHTML, context), source.current = renderedForSource.outerHTML));
  }), story;
};

// src/entry-preview-docs.ts
var decorators = [sourceDecorator], parameters = {
  docs: {
    story: { inline: !0 },
    source: {
      type: SourceType2.DYNAMIC,
      language: "html",
      code: void 0,
      excludeDecorators: void 0
    }
  }
};
export {
  decorators,
  parameters
};
